11-payment-encoding.md

```diff
@@ -140,6 +140,8 @@ Currently defined tagged fields are:
 * `p` (1): `data_length` 52. 256-bit SHA256 payment_hash. Preimage of this provides proof of payment.
 * `s` (16): `data_length` 52. This 256-bit secret prevents forwarding nodes from probing the payment recipient.
 * `d` (13): `data_length` variable. Short description of purpose of payment (UTF-8), e.g. '1 cup of coffee' or 'ナンセンス 1杯'
+* `m` (27): `data_length` variable. Additional metadata to attach to the
+  payment. Note that implementations may apply limits to the size of this field because the size of the hop payload itself is limited as well.
```

Comment:
This isn't clear, and really hard to implement to - if an implementation hits the limit does it fail to send the payment or does it truncate the result? Are implementations expected to simply add a hard-coded limit and fail invoices over a threshold? If so what is that threshold supposed to be (and, critically, below which number should an invoice-creator target keeping this value to ensure their invoices will be parsed?).

Reply:
What I implemented is:
* Fail payment when limit is exceeded, no truncation.
* No hard-coded limit. It tries to find a route that can deliver the full metadata and stay below the 1300 bytes max size. This means that the more metadata there is, the shorter the max route length. This may be a concern, because the receiver can influence pathfinding indirectly.
* Receivers need to be aware that long metadata strings may decrease the success chance for the payment.

Reply:
> If so what is that threshold supposed to be (and, critically, below which number should an invoice-creator target keeping this value to ensure their invoices will be parsed?)

That's a hard question. AFAICT we have only two options:

1. The spec provides a number here (e.g. 64 bytes). It's almost the same as making this field fixed-size, which is the mistake we made with `payment_secret` and don't want to make again. But the encoding is variable-size, so it's better than `payment_secret` and probably easier to update later (if we want to make the size bigger)?
2. The spec doesn't provide a number: every implementation will use a different limit (or every payment will use a different limit with Joost's implementation), which makes it really hard for invoice creators to ensure their invoices can be paid.

I don't think there is a clear winner. Even if invoice creators keep their `payment_metadata` small (below a spec-defined limit), there is always a risk that payers need many hops and are unable to include it in the onion for the (long) paths they find.

Reply:
Limiting `payment_metadata` to something like 64 bytes does sort of defeat the purpose of stateless invoices, unless it is deemed to be sufficient to just include basic invoice data in there for reconstruction.

If you want to put in a delivery address for example, it wouldn't really work.

I would be in favor of letting services decide on the routability/size tradeoff themselves. For custom records, it works in a similar way. The more data is included, the shorter the max route length.

Reply:
Also not quite clear on what the right thing to do is here. Any fixed value we put in here is both limiting for some applications and signaling safety to others ("64 bytes will definitely work" 🤔). It's probably best to mention the risk of not fitting in an onion and defer the decision to the developers using it.

Given the tiny network diameter I personally don't see it increasing the failure rate considerably (6 hops << 20 hops -> ~14*65 bytes of room for random signalling in the onion)

Reply:
I agree its not clear what to do here. I imagine 64 bytes (plus the 32 byte payment secret) should suffice for most use-cases, but if we don't want to hard-code it then I believe the spec should clearly state that nodes MUST NOT apply a limit except as required by routing failures due to too-large-onion.

Relatedly, and maybe invalidating the above, is there a concern of someone detecting the payment distance in a repeated payment by varying the length here and detecting when the payment fails? Maybe we can limit to, like, 128 bytes?

Reply:
Discussed in the spec meeting: allow metadata to use all remaining space in the onion and state that implementations shouldn't limit the size to a lower number.

Reply:
Added